alf/algorithms/vq_vae.py

```diff
@@ -0,0 +1,164 @@
+    2. Standard VAE uses Gaussian prior and posterior. VQVAE can be viewed as
+       using a determinstic form of posterior, which is a categorical
+       distribution with onehot samples computed by nearest neighbor matching
+       (Eq.1 of the paper). By using a uniform prior, the KL divergence is constant.
+
+
+    """
+
+    def __init__(self,
+                 input_tensor_spec,
+                 num_embeddings,
+                 embedding_dim,
+                 encoder_ctor,
+                 decoder_ctor,
+                 optimizer=None,
+                 commitment_loss_weight=1.0,
+                 debug_summaries=False,
+                 name="Vqvae"):
+        """
+        Args:
+            input_tensor_spec (TensorSpec): the tensor spec of
+                the input.
+            num_embeddings (int): the number of embeddings (size of codebook)
+            embedding_dim (int): the dimensionality of embedding vectors
+            encoder_ctor (Callable): called as ``encoder_ctor(observation_spec)``
+                to construct the encoding ``Network``. The network takes raw observation
+                as input and output the latent representation.
+            decoder_ctor (Callable): called as ``decoder_ctor(latent_spec)`` to
+                construct the decoder.
+            optimizer (Optimzer|None): if provided, it will be used to optimize
+                the parameter of encoder_net, decoder_net and embedding vectors.
+            commitment_loss_weight (float): the weight for commitment loss.
+        """
+        super().__init__(debug_summaries=debug_summaries, name=name)
+
+        self._embedding_dim = embedding_dim
+        self._num_embeddings = num_embeddings
+
+        # [n, d]
+        self._embedding = nn.Embedding(self._num_embeddings,
+                                       self._embedding_dim)
+        self._embedding.weight.data.uniform_(-1 / self._num_embeddings,
+                                             1 / self._num_embeddings)
+
+        self._encoding_net = encoder_ctor(input_tensor_spec)
+
+        self._decoding_net = decoder_ctor(self._encoding_net.output_spec)
+
+        if optimizer is not None:
+            self.add_optimizer(
+                optimizer,
+                [self._encoding_net, self._decoding_net, self._embedding])
+        self._optimizer = optimizer
+
+        self._commitment_loss_weight = commitment_loss_weight
+
+    def _predict_step(self, inputs, state=()):
+        """
+        Args:
+            inputs (tensor): with the shape the same as input_tensor_spec
+        """
+        # [B, d]
+        input_embedding, _ = self._encoding_net(inputs)
+
+        # calculate distances
+        # [B, 1] + [n] + [B, n]
+        distances = (
+            torch.sum(input_embedding**2, dim=1, keepdim=True) + torch.sum(
+                self._embedding.weight**2, dim=1) -
+            2 * torch.matmul(input_embedding, self._embedding.weight.t()))
```

Comment:
I feel like this NN isn't as robust as Gumbel-softmax with hard samples.

Reply:
Not sure but it could be